File: android/src/main/kotlin/com/snowplowanalytics/snowplow_flutter_tracker/readers/events/StructuredReader.kt
// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_flutter_tracker.readers.events

import com.snowplowanalytics.snowplow.event.Structured

class StructuredReader(values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val category: String by values
    val action: String by values
    val label: String? by valuesDefault
    val property: String? by valuesDefault
    val value: Double? by valuesDefault

    fun toStructured(): Structured {
        val structured = Structured(category, action)
        label?.let { structured.label(it) }
        property?.let { structured.property(it) }
        value?.let { structured.value(it) }
        return structured
    }
}

File: ios/Classes/readers/events/StructuredReader.swift
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import Foundation
import SnowplowTracker

struct StructuredReader: Decodable {
    let category: String
    let action: String
    let label: String?
    let property: String?
    let value: Double?
}

extension StructuredReader {
    func toStructured() -> Structured {
        let event = Structured(category: category, action: action)
        if let label = self.label { event.label(label) }
        if let prop = self.property { event.property(prop) }
        if let value = self.value { event.value(NSNumber(value: value)) }
        return event
    }
}
